File: Algorithms/Ordered Map/729. My Calendar I.java
class MyCalendar {
    TreeMap<Integer, Integer> events;
    public MyCalendar() {
        events = new TreeMap<Integer, Integer>();
    }
    
    public boolean book(int start, int end) {
        Integer upperBoundKey = events.higherKey(start);
        if(upperBoundKey != null && events.get(upperBoundKey) < end) return false;
        events.put(end, start);
        return true;
    }
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */

File: Algorithms/Ordered Map/729. My Calendar I.cpp
class MyCalendar {
public:
    map<int, int> books;
    
    MyCalendar() {
    }
    
    bool book(int start, int end) {
        map<int, int>::iterator upperIt = books.upper_bound(start);
        if(upperIt != books.end() && upperIt->second < end) return false;
        books[end] = start;
        return true;
    }
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */
